scio-core/src/main/scala/com/spotify/scio/util/ScioUtil.scala

```diff
@@ -115,12 +115,16 @@ private[scio] object ScioUtil {
     DefaultFilenamePolicy.fromStandardParameters(prefix, shardTemplate, suffix, isWindowed)
   }
 
-  def tempDirOrDefault(tempDirectory: String, sc: ScioContext): ResourceId = {
-    Option(tempDirectory).map(toResourceId).getOrElse {
-      val tempLocationOpt: String = sc.options.getTempLocation
-      FileSystems.matchNewResource(tempLocationOpt, true)
-    }
-  }
+  def tempDirOrDefault(tempDirectory: String, sc: ScioContext): ResourceId =
+    Option(tempDirectory)
+      .map(toResourceId)
+      .orElse(
+        Try(sc.optionsAs[GcpOptions]).toOption
+          .flatMap(gcpOptions => Option(gcpOptions.getGcpTempLocation))
+          .orElse(Option(sc.options.getTempLocation))
+          .map(FileSystems.matchNewResource(_, true))
+      )
+      .getOrElse(throw new IllegalStateException("Unable to find any temp location"))
```

Comment:
We can re-use `toResourceId`. I think we'd also prefer to use `tempLocation` over `gcpTempLocation`?

```suggestion
  def tempDirOrDefault(tempDirectory: String, sc: ScioContext): ResourceId = {
    Option(tempDirectory)
      .orElse(Option(sc.options.getTempLocation))
      .orElse(Try(sc.optionsAs[GcpOptions]).toOption.flatMap(x => Option(x.getGcpTempLocation)))
      .map(toResourceId)
      .getOrElse(throw new IllegalStateException("Unable to find any temp location"))
  }
```